File: Java/Graphs/ProblemsOnDfsBfs/1036-rotting-oranges/rotting-oranges.java
class Pair{
    int row;
    int col;
    int time;
    public Pair(int row, int col, int time){
        this.row = row;
        this.col = col;
        this.time = time;
    }
}
class Solution {
    public int orangesRotting(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        Queue<Pair> q = new LinkedList<>();

        int visited[][] = new int[n][m];
        int countFresh = 0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j] == 2){
                    q.add(new Pair(i,j,0));
                    visited[i][j] = 2;
                }else{
                    visited[i][j] = 0;
                }
                
                if(grid[i][j] == 1)countFresh++;
            }
        }

        int time=0;
        int di[] = {-1,0,1,0};
        int dj[] = {0,1,0,-1};
        int cnt=0;
        while(!q.isEmpty()){
            int r = q.peek().row;
            int c = q.peek().col;
            int t = q.peek().time;
            time = Math.max(t,time);
            q.remove();
            for(int i=0;i<4;i++){
                int nrow = r + di[i];
                int ncol = c + dj[i];
                if(nrow>=0 && nrow<n && ncol>=0 && ncol<m && visited[nrow][ncol] == 0 && grid[nrow][ncol] == 1){
                    q.add(new Pair(nrow,ncol, t+1));
                    visited[nrow][ncol] = 2;
                    cnt++;
                }
            }
        }
        if(cnt != countFresh)return -1;
        return time;
    }
}

File: Python/Graphs/ProblemsOnBFSAndDFS/1036-rotting-oranges/rotting-oranges.py
class Pair:
    def __init__(self, row:int, col:int,time:int):
        self.row = row
        self.col = col
        self.time = time

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        n = len(grid)
        m = len(grid[0])
        q = deque()

        fresh_count = 0        
        for row in range(n):
            for col in range(m):
                if grid[row][col] == 2:
                    q.append(Pair(row,col,0))
                if grid[row][col] == 1:
                    fresh_count+=1

        delRow = [-1,0,1,0]
        delCol = [0,1,0,-1]

        q_cnt=0
        timing=0
        while q:
            row = q[0].row
            col = q[0].col
            time = q[0].time
            timing= max(timing,time)
            q.popleft()
            for i in range(4):
                n_row = row + delRow[i]
                n_col = col + delCol[i]

                if(n_row >= 0 and n_row < n and n_col >=0 and n_col <m):
                    if(grid[n_row][n_col] == 1):
                        q_cnt+=1
                        grid[n_row][n_col] = 2
                        q.append(Pair(n_row,n_col,time+1))
        
        if q_cnt == fresh_count:
            return timing
        else:
            return -1